### ISSUE TITLE
Expose HAST nodes to components

### ISSUE THREAD

**Issue by Rokt33r** [rehypejs/rehype-react#19]
### Subject of the feature

I want to access HAST nodes from React components to use position or custom attributes of nodes when rendering. 

### Problem

For now, there is no way to access because rehype-react is tightly bound to hast-to-hyperscript, which consumes Node.

### Expected behavior

A React component should be able to access its Node info somehow.

### Alternatives

Passing a node instance via `node` prop.

**Comment by wooorm** [comment]
What would this look like? Do you get the whole subtree of the node that is to be compiled? Can you access `children` of a node too (which are compiled before the parent?)

The simplest solution I can think of to begin solving it is:

https://github.com/syntax-tree/hast-to-hyperscript/blob/6337edb309209ba2ec8c2431daa779f50d2a7ad2/index.js#L141:

```diff
  // Ensure no React warnings are triggered for void elements having children
  // passed in.
  result =
-   elements.length === 0 ? h(name, attributes) : h(name, attributes, elements)
+   elements.length === 0 ? h.call(node, name, attributes) : h.call(node, name, attributes, elements)

  // Restore parent schema.
  ctx.schema = parentSchema
```

...as passing it in props will probably cause things to fail because some tools would serialize `node` as an attribute?

...then we have access to it here:

https://github.com/rehypejs/rehype-react/blob/761597e38a9941e4d99d95c75262710dec09a912/index.js#L39-L43

...and can choose to put it in a prop for components?

**Comment by Rokt33r** [comment]
> What would this look like? Do you get the whole subtree of the node that is to be compiled? Can you access children of a node too (which are compiled before the parent?)

For now, I just need to access the node being compiled. But I guess some users might want to access whole subtree too.

> The simplest solution I can think of to begin solving it is: ...

The solution looks good to me. Is there any way to modify the `h` function from rehype-react? If it doesn't exist, I guess I should fork this package and implement customized `h` function.

**Comment by wooorm** [comment]
The first example shows how `h.call` is added to `hast-to-hyperscript`, which exposes the node that is transformed in `rehype-react`’s `h` as `this`.

We could then change `rehype-react` to something like this:

```js
 function h(name, props, children) { 
   var component = name
   if (has.call(components, name)) {
     component = components[name]
     props.node = this
   }
   return createElement(component, props, children) 
 } 
```

**Comment by Rokt33r** [comment]
LGTM. And I think we should make it optional too for users who are using [prop-types package](https://www.npmjs.com/package/prop-types). So I guess I should submit a pr to hast-to-hyperscript first?

**Comment by wooorm** [comment]
PR to h2h is a good first step, I don’t see a harm in changing that.

I do feel like an option to pass `node` to *every* component is a bit weird. 🤔 How would this (whether to pass a prop or not) normally go in React-land?

**Comment by Rokt33r** [comment]
> I do feel like an option to pass node to every component is a bit weird

How is it weird?

**Comment by wooorm** [comment]
I mean, a single option `passNode` or so that passes `props.node` to every component, is that powerful enough? Or will people then want an option `passNodeTo: ['ThisComponent', 'ThatComponent']`? Will people also want to access `parent` in props potentially?

**Comment by Rokt33r** [comment]
> I mean, a single option passNode or so that passes props.node to every component, is that powerful enough?

I think this sounds enough in my opinion.

> Will people also want to access parent in props potentially?

I think this can be done while transforming. So I don't think we need it.

**Comment by wooorm** [comment]
Alright, sounds like we have a plan! Are you interested / able to work on it?

**Comment by Rokt33r** [comment]
Yeah, I'll do it tonight or tomorrow!

**Comment by wooorm** [comment]
So this is 50% there. Last part is to update here, right?

**Comment by wooorm** [comment]
@Rokt33r No rush at all, but a friendly ping!

**Comment by Rokt33r** [comment]
Ah sorry. I'll do it now.

**Comment by wooorm** [comment]
Released in 6.1.0!
